src/main/java/org/takes/rs/RsWithHeader.java

```diff
@@ -94,22 +95,66 @@ public RsWithHeader(final Response res, final CharSequence name,
      */
     public RsWithHeader(final Response res, final CharSequence header) {
         super(
-            new Response() {
-                @Override
-                public List<String> head() throws IOException {
-                    final List<String> head = new LinkedList<String>();
-                    for (final String hdr : res.head()) {
-                        head.add(hdr);
-                    }
-                    head.add(header.toString());
-                    return head;
-                }
-                @Override
-                public InputStream body() throws IOException {
-                    return res.body();
-                }
-            }
+            new DecoratedResponse(res, header)
         );
     }
 
+    /**
+     * Response with an additional header
+     *
+     * <p>The class is immutable and thread-safe.
+     */
+    private static final class DecoratedResponse implements Response {
+        /**
+         * Pattern for all other lines in the head.
+         */
+        private static final Pattern HEADER = Pattern.compile(
+            "[a-zA-Z\\-]+:\\p{Print}+"
+        );
+
+        /**
+         * Original response.
+         */
+        private final transient Response response;
+
+        /**
+         * Header to add.
+         */
+        private final transient CharSequence header;
+
+        /**
+         * Ctor.
+         * @param res Original response
+         * @param extension Header to add
+         */
+        DecoratedResponse(final Response res, final CharSequence extension) {
```

Comment:
@YamStranger lets rename `extension` to `hdr` according to http://www.yegor256.com/2014/04/27/typical-mistakes-in-java-code.html#variable-names
> Sometimes, you may have collisions between constructor parameters and in-class properties if the constructor saves incoming data in an instantiated object. In this case, I recommend to create abbreviations by removing vowels (see how USPS abbreviates street names).